### PR TITLE
Correctly handle resource variables in control_dependencies

### PR DESCRIPTION
**Pull request by lgeiger**
This PR includes the `FuncGraph` changes required to rollforward #40564 and fixes handling of resource variables as input to control dependencies.

/cc @alextp @crccw

### CODE REVIEW

File: tensorflow/python/framework/ops.py [hunk 1]
```diff
@@ -4737,6 +4737,10 @@ def my_func(pred, tensor):
     """
     if control_inputs is None:
       return self._ControlDependenciesController(self, None)
+
+    # Imported here to avoid circular dependency.
+    from tensorflow.python.ops import resource_variable_ops  # pylint: disable=g-import-not-at-top
```
Comment (alextp) [inline tensorflow/python/framework/ops.py:4742]: I think this circular import can fail in some internal google things. Let's try though.

**Review by alextp — APPROVED** [review]

File: tensorflow/python/framework/ops.py (hunk 1 above)
Comment (lgeiger) [reply tensorflow/python/framework/ops.py:4742]: For the `AutoCastVariable` use case only the func_graph changes are strictly necessary. I only included them here for consistency, but I am happy to revert which would fix any circular dependencies.

Comment (alextp) [reply tensorflow/python/framework/ops.py:4742]: Let's see what our CI says. If it complains we can revert it.

**Comment by lgeiger** [comment]
Looks like the `cla/google` bot is stuck (Just to confirm that I have signed the CLI and have contributed to TF in the past).

**Review by alextp — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

File: tensorflow/python/framework/ops.py (hunk 1 above)
Comment (jaingaurav) [reply tensorflow/python/framework/ops.py:4742]: Could we have further explanation of why the `hasattr` check was not sufficient?

**Review by jaingaurav — CHANGES_REQUESTED** [review]
Can we please add some tests to this change? There seems to a be quite a few subtle changes here which are very hard to understand as to why they are needed.

File: tensorflow/python/framework/func_graph.py [hunk 2]
```diff
@@ -187,7 +187,7 @@ def __init__(self, name, collections=None, capture_by_value=None):
     self.inputs = []
     self.outputs = []
     self.control_outputs = []
-    self.control_captures = set()
+    self.control_captures = object_identity.ObjectIdentitySet()
```
Comment (jaingaurav) [inline tensorflow/python/framework/func_graph.py:190]: Why is this change needed? We try to avoid using `ObjectIdentitySet` internally due to the performance overhead, so if possible it would be best to avoid using it.

**Comment by lgeiger** [comment]
> Can we please add some tests to this change? There seems to a be quite a few subtle changes here which are very hard to understand as to why they are needed.

@jaingaurav That's a fair point. Those changes are mainly required for fixing the `AutoCastVariable` issues when assigning to a variable in a distributed strategy and have been discussed in #40564 which was reverted due to internal failures on TPU.

Unfortunately the only way I was able to find a unittest for this was with the upcoming changes to `AutoCastVariable`. I resubmitted the changes in #41214 which includes [a unittest which would fail without these changes](https://github.com/tensorflow/tensorflow/pull/41214/files#diff-7c464547ccc6362ecbb81b73b4cfbad0R364-R380) and additional comments with links to the relevant discussion that try to answer the the comments you made.
